salt/states/file.py

```diff
@@ -663,6 +663,11 @@ def _clean_dir(root, keep, exclude_pat):
     Clean out all of the files and directories in a directory (root) while
     preserving the files in a list (keep) and part of exclude_pat
     """
+    # Create a case-sensitive dict before doing comparisons for Windows
+    win_keep = None
+    if salt.utils.platform.is_windows():
```

Comment:
are we certain this only impacts windows? From the issues it seems to impact MacOSX or possibly even different file systems that treat different files as case insensitive.

Reply:
i re-read over the ticket and someone with macOS did report the behavior. I'll have to address that in this PR

Reply:
Its not just MacOS, there are other file systems even on linux or other OSs that may run into this as well. see comment (https://github.com/saltstack/salt/issues/47969#issuecomment-394748175)

Reply:
ah yes... perhaps removing the system checks them and just performing this operation regardless of OS/file system is what needs to happen. because there's probably an unknown number of file systems that would experience this unwanted behavior.

Reply:
Would it be appropriate to just _always_ perform this check?

Reply:
I don't think we should always perform the check, because this loop could add extra time that is not necessary for all filesystems. 

1. I'm leaning towards adding a salt.utils check that checks if the filesystem is casesensitive or not. Possibly create a temp file with an uppercase name and check if the operating system detects that it exists when the filename is converted to lowercase.

2. Or default to this behavior for windows and mac, and add an argument so others can see this behavior if they pass in an extra argument.

ping @s0undt3ch or @waynew do either of you have opinions here?

Reply:
Option 1 should take care of this and any other filesystem. My only problem with it is that salt doesn't "remember" that information, so we would be creating temfiles at least once per process ... (Create a grain for this?)
2. seems like the least intrusive approach, but it might leave other filesystems out....

I'm leaning towards 1, perhaps a grain, since we can cache grains....

Reply:
It sounds like then setting a grain on startup and then having a salt.util check for that grain when this function is used is the direction. What about something like this to check?
```
import os.path
home = os.path.expanduser('~')
sensitivity = os.path.exists(home.upper()) and os.path.exists(home.lower())
```
or
```
import os.path
import tempfile

with tempfile.NamedTemporaryFile(prefix='TmP',dir=path, delete=True) as tmp_file:
    return(not os.path.exists(tmp_file.name.lower()))
```

Reply:
I don't think we need a salt.utils in addition to a grain since all we need to do is check if its in the grains dictionary. I'm not sure option one would work for all use cases. Wouldn't you be relying on the home path to have an uppercase in it and not all OSs do? I'm leaning towards option 2 using tempfile.

Reply:
Not necessarily, it is whatever the home path is and python will evaluate whether or not it recognizes if, for example, 
'/HOME/USERNAME' is equivalent to '/home/username' based on the system it is running on. On ubuntu, option 1 returns False and on windows and mac it returns true, at least on my journaled format it does.

Personally i'm good with either, i think both are good options

Reply:
ping @waynew or @s0undt3ch what do you think? Which option?

Reply:
`tempfile` looks like a good approach

Reply:
sounds good, will implement. as for the grain, what should we call it? `case_sensitivity`? `case_sensitive_system`?

Reply:
We probably don't need the grain, a utility would be enough?

```python
def case_sensitive_filesystem(path):
    with tempfile.NamedTemporaryFile(prefix='TmP',dir=path, delete=True) as tmp_file:
        return os.path.exists(tmp_file.name.lower())
```

Reply:
Specially because, at least on Linux, there can be multiple mountpoints with multiple file systems...

Reply:
You recommended the grain approach earlier in the conversation "My only problem with it is that salt doesn't "remember" that information, so we would be creating temfiles at least once per process ... (Create a grain for this?)"

Is that not a concern anymore?

Reply:
yeah, I did, but we don't easily have access to grains on utils, unless we load utils under `__utils__` and we should be avoiding that. Might I say, we should be removing `__utils__` usage as much as possible.

Reply:
Fair point. K ill add those changes :)